lib/vnet/setup.go

```diff
@@ -0,0 +1,150 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package vnet
+
+import (
+	"context"
+	"log/slog"
+	"os"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+	"golang.zx2c4.com/wireguard/tun"
+)
+
+// Run is a blocking call to create and start Teleport VNet.
+func Run(ctx context.Context) error {
+	ipv6Prefix, err := IPv6Prefix()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tun, err := CreateAndSetupTUNDevice(ctx, ipv6Prefix.String())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	manager, err := NewManager(&Config{
+		TUNDevice:  tun,
+		IPv6Prefix: ipv6Prefix,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	allErrors := make(chan error, 3)
+	g, ctx := errgroup.WithContext(ctx)
+	g.Go(func() error {
+		// Make sure to cancel the context if manager.Run terminates for any reason.
+		defer cancel()
+		err := trace.Wrap(manager.Run(ctx))
+		allErrors <- err
+		return err
+	})
+	g.Go(func() error {
+		// Wait until the context is canceled, either from receiving SIGTERM or from manager.Run terminating.
+		<-ctx.Done()
+
+		tunErr := trace.Wrap(tun.Close())
+		allErrors <- tunErr
+
+		destroyErr := trace.Wrap(manager.Destroy())
+		allErrors <- destroyErr
+
+		return trace.NewAggregate(tunErr, destroyErr)
+	})
+	// Deliberately ignoring the error from g.Wait() to return an aggregate of all errors.
+	_ = g.Wait()
+	close(allErrors)
+	return trace.NewAggregateFromChannel(allErrors, context.Background())
+}
```

Comment:
I believe I see a race condition here:

If the first `go.Go` errors, `_ = g.Wait()` will return and then we'll `close(allErrors)`.

Concurrently, the second `g.Go`'s `<-ctx.Done()` will execute, and then we will try to write to the `allErrors` channel. However, if the `close(allErrors)` runs first, writing to this closed channel will cause a panic.

Double check my work, but unless I'm mistaken, I don't think we actually need separate goroutines here. As written, the second goroutine is waiting for the first goroutine to run the cleanup methods, and then we want to wait for both of those to finish before returning. Which is a long winded way of saying that all of this should be done sequentially in a single goroutine, like:

```suggestion
// Run is a blocking call to create and start Teleport VNet.
func Run(ctx context.Context) error {
	ipv6Prefix, err := IPv6Prefix()
	if err != nil {
		return trace.Wrap(err)
	}

	tun, err := CreateAndSetupTUNDevice(ctx, ipv6Prefix.String())
	if err != nil {
		return trace.Wrap(err)
	}

	manager, err := NewManager(&Config{
		TUNDevice:  tun,
		IPv6Prefix: ipv6Prefix,
	})
	if err != nil {
		return trace.Wrap(err)
	}

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	// Wait for the manager to finish running.
	managerErr := trace.Wrap(manager.Run(ctx))

	// Cleanup.
	tunCloseErr := trace.Wrap(tun.Close())
	managerDestroyErr := trace.Wrap(manager.Destroy())

	// Return aggregate error.
	return trace.NewAggregate(managerErr, tunCloseErr, managerDestroyErr)
}
```

Reply:
g.Wait() blocks until all function calls from the Go method have returned, so the race isn't a concern. Also, I want to concurrently close the tun and destroy the manager when the context is cancelled, in case manager.Run doesn't quit on its own, which it might not

Reply:
> g.Wait() blocks until all function calls from the Go method have returned, so the race isn't a concern.

True, I had a misunderstanding of how `g.Wait()` works

> Also, I want to concurrently close the tun and destroy the manager when the context is cancelled, in case manager.Run doesn't quit on its own, which it might not

If `manager.Run` doesn't quit on it's own when `ctx.Done()` receives, is that not an abuse of `context.Context`? My general understanding is that if `ctx` is passed to a function, there's an implied contract function should return when that `ctx` is cancelled.

If that interpretation is correct, then consider encapsulating the cleanup logic into the `m.Run()` function itself. E.g. make this function look like

```go
// Run is a blocking call to create and start Teleport VNet.
func Run(ctx context.Context) error {
	ipv6Prefix, err := IPv6Prefix()
	if err != nil {
		return trace.Wrap(err)
	}

	tun, err := CreateAndSetupTUNDevice(ctx, ipv6Prefix.String())
	if err != nil {
		return trace.Wrap(err)
	}

	manager, err := NewManager(&Config{
		TUNDevice:  tun,
		IPv6Prefix: ipv6Prefix,
	})
	if err != nil {
		return trace.Wrap(err)
	}

	// Return aggregate error.
	return trace.Wrap(manager.Run(ctx))
}
```

and then modify `Manager`'s `Run` and `Destroy` to take care of the context contract

```go
// Run starts the VNet.
func (m *Manager) Run(ctx context.Context) error {
	m.slog.InfoContext(ctx, "Running Teleport VNet.", "ipv6_prefix", m.ipv6Prefix)
	errs := make(chan error, 2)
	ctx, cancel := context.WithCancel(ctx)
	go func() {
		defer cancel()
		errs <- m.statsHandler(ctx)
	}()
	go func() {
		defer cancel()
		errs <- forwardBetweenTunAndNetstack(ctx, m.tun, m.linkEndpoint)
	}()

	// Wait for either of the goroutines to return an error.
	<-ctx.Done()
	// Ensures that forwardBetweenTunAndNetstack goroutine returns.
	destroyErr := m.Destroy()
	// Aggregate any errors that were returned from the goroutines.
	groutineErrs := trace.NewAggregateFromChannel(errs, context.Background())
	// Return an aggregate error.
	return trace.Wrap(trace.NewAggregate(groutineErrs, destroyErr))
}

// Destroy closes the link endpoint, waits for all goroutines to terminate, and destroys the networking stack.
func (m *Manager) Destroy() error {
	close(m.destroyed)
	m.linkEndpoint.Close()
	m.wg.Wait()
	m.stack.Destroy()
	return trace.Wrap(m.tun.Close()) // Close the TUN device, which will stop the forwarder goroutines.
}
```

Reply:
alright, check out the latest, I got rid of Destroy and Manager just reliably quits when the context is cancelled now